tools/distpartitioning/data_shuffle.py

```diff
@@ -179,103 +204,106 @@ def exchange_node_features(rank, world_size, node_data, node_features, ntypes_ma
+        #as specified with this range. So no. of rows = ending-idx - starting-idx.
+        for feat_info in ntype_info:
+
+            #determine the owner process for these node features. 
+            node_feats_per_rank = []
+            global_nid_per_rank = []
+            feat_name = feat_info[0]
+            feat_key = ntype_name+'/'+feat_name
+
+            #compute the global_nid range for this node features
+            type_nid_start = int(feat_info[1])
+            type_nid_end = int(feat_info[2])
+            begin_global_nid = ntype_gnid_map[ntype_name][0]
+            gnid_start = begin_global_nid + type_nid_start
+            gnid_end = begin_global_nid + type_nid_end
+
+            #type_nids for this feature subset on the current rank
+            gnids_feat = np.arange(gnid_start, gnid_end)
+            tnids_feat = np.arange(type_nid_start, type_nid_end)
+            local_idx = np.arange(0, type_nid_end - type_nid_start)
+
+            #check if node features exist for this ntype_name + feat_name
+            #this check should always pass, because node_feature_tids are built
+            #by reading the input metadata json file for existing node features.
+            assert(feat_key in node_features)
+
+            node_feats = node_features[feat_key]
+            for part_id in range(world_size):
+                partid_slice = node_part_ids[gnid_start:gnid_end]
+                cond = (partid_slice == part_id)
+                gnids_per_partid = gnids_feat[cond]
+                tnids_per_partid = tnids_feat[cond]
+                local_idx_partid = local_idx[cond]
+
+                if (gnids_per_partid.shape[0] == 0):
+                    node_feats_per_rank.append({feat_key : torch.empty((0,), dtype=torch.float)})
+                    global_nid_per_rank.append({feat_key : torch.empty((0,), dtype=torch.int64)})
+                else:
+                    node_feats_per_rank.append({feat_key : node_feats[local_idx_partid]})
+                    global_nid_per_rank.append({feat_key : gnids_per_partid})
+
+            #features (and global nids) per rank to be sent out are ready
+            #for transmission, perform alltoallv here.
+            output_feat_list = alltoall_cpu_object_lst(rank, world_size, node_feats_per_rank)
+            output_feat_list[rank] = node_feats_per_rank[rank]
+
+            output_nid_list = alltoall_cpu_object_lst(rank, world_size, global_nid_per_rank)
+            output_nid_list[rank] = global_nid_per_rank[rank]
+
+            #stitch node_features together to form one large feature tensor
+            for idx, x in enumerate(output_feat_list):
+                if feat_key in own_node_features:
+                    t = own_node_features[feat_key]
+                    own_node_features[feat_key] = torch.cat([t, x[feat_key]])
+                    t = own_global_nids[feat_key]
+                    own_global_nids[feat_key] = np.concatenate([t, output_nid_list[idx][feat_key]])
```

Comment:
the same comments as above. the code here should be simplified.
we should keep all tensors in the list and call concatenate once, instead of calling it in every iteration.

Reply:
Changed the code appropriately. Will also search for this kind of pattern in the entire code base and change accordingly.